pkg/controller/deployment/failed.go

```diff
@@ -0,0 +1,170 @@
+// and when new pods scale up or old pods scale down. Progress is not estimated for paused
+// deployments, or when the deployment needs to rollback, or when users don't really care
+// about it ie. progressDeadlineSeconds is not specified.
+func (dc *DeploymentController) hasTimedOut(d *extensions.Deployment) (bool, error) {
+	if d.Spec.ProgressDeadlineSeconds == nil || d.Spec.RollbackTo != nil || d.Spec.Paused {
+		return false, nil
+	}
+
+	newRS, oldRSs, err := dc.getAllReplicaSetsAndSyncRevision(d, false)
+	if err != nil {
+		return false, err
+	}
+	allRSs := oldRSs
+
+	// There is a template change so we don't need to check for any progress right now.
+	if newRS == nil {
+		return false, nil
+	}
+
+	// Look at the status of the deployment; if there is already a NewRSAvailableReason
+	// then we don't need to estimate any progress. This is needed in order to avoid
+	// estimating progress for scaling events after a rollout has finished.
+	cond := util.GetCondition(d.Status, extensions.DeploymentProgressing, api.ConditionTrue)
+	if cond != nil && cond.Reason == util.NewRSAvailableReason {
+		return false, nil
+	}
+
+	allRSs = append(allRSs, newRS)
+	newStatus, err := dc.calculateStatus(allRSs, newRS, d)
+	if err != nil {
+		return false, err
+	}
+
+	if util.IsDeploymentComplete(d, newStatus) ||
+		util.IsDeploymentProgressing(d, newStatus) {
+		return false, nil
+	}
+
+	return util.IsDeploymentFailed(d, newStatus), nil
+}
+
+func (dc *DeploymentController) syncFailed(deployment *extensions.Deployment) error {
+	newRS, oldRSs, err := dc.getAllReplicaSetsAndSyncRevision(deployment, false)
+	if err != nil {
+		return err
+	}
+
+	allRSs := append(oldRSs, newRS)
+	return dc.syncRolloutStatus(allRSs, newRS, deployment)
+}
+
+// syncRolloutStatus updates the status of a deployment during a rollout or
+// when it has failed progressing.
+func (dc *DeploymentController) syncRolloutStatus(allRSs []*extensions.ReplicaSet, newRS *extensions.ReplicaSet, d *extensions.Deployment) error {
+	newStatus, err := dc.calculateStatus(allRSs, newRS, d)
+	if err != nil {
+		return err
+	}
+
+	if d.Spec.ProgressDeadlineSeconds != nil {
+		switch {
+		case util.IsDeploymentComplete(d, newStatus):
+			// Update the deployment conditions with a message for the new replica set that
+			// was successfully deployed. If the condition already exists, ignore this update.
+			// Cleanup any condition that reports lack of progress.
+			msg := fmt.Sprintf("Replica set %q has successfully progressed.", newRS.Name)
+			condition := util.NewCondition(extensions.DeploymentProgressing, api.ConditionTrue, util.NewRSAvailableReason, msg)
+			if !util.ConditionExists(newStatus, *condition) {
+				newStatus = util.SetCondition(newStatus, *condition)
+			}
+			// Cleanup conditions that denote any failures.
+			newStatus, _ = util.RemoveCondition(newStatus, extensions.DeploymentProgressing, api.ConditionFalse)
```

Comment:
So that users won't be confused and start asking why their deployments while healthy have all these failure conditions. We have seen it already in DeploymentConfigs when we were emitting events for update conflicts.